### DISCUSSION TITLE
More flexible module naming

### DISCUSSION THREAD

**Discussion by markagama** [oqtane/oqtane.framework#3596] [Ideas]
I would like to be able to name modules using the schema: Agama.Oqtane.Module.Mymodule
If I put agama.oqtane in the owner name field I get an error message.

I cannot get Agama.Module.Oqtane.Mymodule because of the same error - no punctuation
And as this is not the schema I use with other CMS, this would not be a preferred solution

It would be really helpful if this could be accommodated in the next update.

**Comment by thabaum** [comment]
Hello @markagama 

This issue is being discussed further at issue #3597

The reason is "oqtane" cannot be used due to namespace conflicts that causes issues within the Oqtane Framework which was explained in the issue linked above.

@sbwalker said there maybe a way to allow it if it does not start with "Oqtane" and needs tested to ensure no conflicts.  The discussion is currently in this issue created.

Hopefully this can be handled to be a little more flexible to where it works for your desired namespaces.  Please add any feedback in that issue to help if this is something you wish to proceed now you know the reason why it wont allow this.

Cheers!

**Reply by markagama** [reply]
thank you @thabaum for the help with this matter.

As mentioned at the top, my normal approach to creating new modules has been:

companyname.cmsname.modulename.

e.g. agama.oqtane.module.name <- the inclusion of "module" appears to be mandatory in Oqtane

Anything that goes partway towards this approach is helpful.

I also queried why a GUID is not used to discern between modules.  Without a GUID it becomes possible to use the same module name as that created by someone else. Is this a security problem that should be considered? Will modules be registered to prevent module hijacking? There appears to be a lot of this going on in the microsoft / android space. Possibly this is a concern that is being left for later, but in today's world security should be a foundation of all software projects.

**Reply by sbwalker** [reply]
@markagama please note that although the Module Creator contains certain patterns and validation rules for scaffolding new modules, it does not prevent you from modifying the code once the module is created. So if you want to use a specific namespace convention, you can easily modify it in your favorite code editor (ie. search and replace). That being said, we will investigate adjusting the validation rules to allow for more flexibility in the Module Creator as well.

In regards to the suggestion about using a GUID, it would not solve any perceived naming collision problems. This is because module identifiers are created by developers and are embedded in the source code. So if you used GUID 123 in your module and I decided to use GUID 123 as well, there is nothing to prevent it from happening. So a GUID offers no benefit over a name in this scenario.  However, since a name can communicate context it is actually the better approach.  If a module is intended to be shared with others as either an open source or commercial extension, the expectation is that it will be registered in the Marketplace (https://www.oqtane.net) which will ensure that the name is unique and is not being used by multiple developers.
